File: app/src/main/java/com/example/weeboos/askforpermissionlib/MainActivity.java
package com.example.weeboos.askforpermissionlib;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import com.example.weeboos.permissionlib.PermissionRequest;
import com.example.weeboos.permissionlib.PermissionUtils;

import java.util.ArrayList;


public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        final String[] permissionNames = new String[]{
                Manifest.permission.INSTALL_PACKAGES,
        };
        findViewById(R.id.tv_ask).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                PermissionRequest.getInstance().build(MainActivity.this).requestPermission(new PermissionRequest.PermissionListener() {
                    @Override
                    public void permissionGranted() {
                        Toast.makeText(MainActivity.this,"获取成功",Toast.LENGTH_SHORT).show();
                    }

                    @Override
                    public void permissionDenied(ArrayList<String> permissions) {
                        Toast.makeText(MainActivity.this,"获取拒绝",Toast.LENGTH_SHORT).show();
                    }

                    @Override
                    public void permissionNeverAsk(ArrayList<String> permissions) {
                        Toast.makeText(MainActivity.this,"不再询问",Toast.LENGTH_SHORT).show();
                        PermissionUtils.showAlertDialog(MainActivity.this,PermissionUtils.translateArrayString(permissions));
                    }
                },permissionNames);
            }
        });
    }


}

File: app/src/main/java/com/example/weeboos/askforpermissionlib/MainActivity.kt
package com.example.weeboos.askforpermissionlib

import android.Manifest
import android.annotation.TargetApi
import android.app.Application
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.provider.Settings
import android.support.annotation.StringDef
import android.widget.Toast
import com.example.weeboos.permissionlib.PermissionPool
import com.example.weeboos.permissionlib.PermissionRequest

class MainActivity : AppCompatActivity() {

    @TargetApi(Build.VERSION_CODES.M)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        PermissionRequest(this).requestPermission(object:PermissionRequest.PermissionListener{
            override fun permissionGranted(permission: String) {
                Toast.makeText(this@MainActivity,"拥有权限",Toast.LENGTH_LONG).show()
                //获取到权限，进行下一步
            }

            override fun permissionDenied(permission: String) {
                requestPermissions(arrayOf(permission), PermissionPool.CAMERA)
                Toast.makeText(this@MainActivity,"未拥有权限",Toast.LENGTH_LONG).show()
            }

            override fun permissionNeverAsk(permission: String) {
                Toast.makeText(this@MainActivity,"已拒绝再次询问",Toast.LENGTH_LONG).show()
                //跳转设置页面设置权限
                goToSetting()
            }
        }, Manifest.permission.CAMERA)
    }

    /**
     * 跳转前往设置
     */
    fun goToSetting() {
        val intent = Intent()
        intent.action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
        val uri = Uri.fromParts("package", packageName, null)
        intent.data = uri
        startActivity(intent)
    }
}
